functions/send_encounters/main.py

```diff
@@ -0,0 +1,123 @@
+import logging
+import os
+import secrets
+from datetime import datetime
+from typing import Optional
+
+from flask import jsonify
+from google.cloud import bigquery, datastore
+from google.cloud.datastore import Entity
+
+BQ_TABLE_ID = f"{os.environ['GCP_PROJECT']}.{os.environ['BQ_DATASET']}.{os.environ['BQ_TABLE']}"
+
+datastore_client = datastore.Client()
+
+
+def send_encounters(request):
+    if not request.is_json:
+        return jsonify(
+            {'status': 'failed',
+             'message': 'invalid data'
+             }
+        ), 422
+
+    request_data = request.get_json()
+
+    for key in ["user_id", "platform", "os_version", "device_type", "app_version", "lang", "encounters"]:
+        if key not in request_data:
+            return jsonify(
+                {
+                    'status': 'failed',
+                    'message': f'missing field: {key}',
+                }
+            ), 422
+
+    user_id = request_data["user_id"]
+    platform = request_data["platform"]
+    os_version = request_data["os_version"]
+    device_type = request_data["device_type"]
+    app_version = request_data["app_version"]
+    lang = request_data["lang"]
+    encounters = request_data["encounters"]
+
+    user_entity = _get_user_entity(user_id)
+    upload_id = secrets.token_hex(32)
+    if not user_entity:
+        return jsonify(
+            {
+                'status': 'failed',
+                'message': f'unauthorized',
+            }
+        ), 401
+    rows_to_insert = []
+    for encounter in encounters:
+        for key in ["encounter_date", "beacon_id", "signal_strength"]:
```

Comment:
Could move all validation to functions. It is better to keep the function as clear as possible.